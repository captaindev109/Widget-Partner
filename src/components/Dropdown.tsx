import Card from '@/generics/Card';
import Label from '@/generics/Label';
import styled from 'styled-components';
import Tooltip from './Tooltip';

import { CreateSVGComponent } from '@/components/SVGIcons/Index';
import Text from '@/generics/Text';
import { animateOpacity, ANIMATION_DURATION } from '@/utils/animationUtils';
import { BasicString, BasicBool, BasicAny } from '@/types/types';
import { FC, useEffect, useRef } from 'react';

export interface DropdownOption {
  label: string;
  value: string;
}

export interface DropdownProps {
  options: DropdownOption[];
  onChange?: (option: DropdownOption) => void;
  value?: BasicString;
  disabled?: BasicBool;
  placeholder?: BasicString;
  label?: BasicString;
  name?: BasicString;
  tooltip?: BasicString;
  helperText?: BasicString;
  error?: BasicBool;
  icon?: BasicBool;
  register?: BasicAny;
}

export interface DropdownOption {
  label: string;
  value: string;
}

const TopRow = styled(Card)`
  display: flex;
  align-items: center;

  & > * {
    margin-left: 10px;
  }

  margin-bottom: 8px;
`;

const StyledIcon = styled(Card)<{ $animate: boolean; error?: boolean }>`
  color: ${(props) => (props.error ? props.theme.color.red600 : props.theme.color.gray700)};
  transition: all ${ANIMATION_DURATION} ease;

  ${(props) => props.$animate && animateOpacity()}
`;

const HelperText = styled(Text)<{ $animate: boolean; error?: boolean }>`
  color: ${(props) => (props.error ? props.theme.color.red600 : props.theme.color.gray700)};
  transition: color ${ANIMATION_DURATION} ease;

  ${(props) => props.$animate && animateOpacity()}
`;

const IconWrapper = styled(Card)`
  position: absolute;
  top: 0;
  bottom: 0;
  right: 24px;

  display: flex;
  align-items: center;
  pointer-events: none;
`;

const StyledSelect = styled.select<BasicAny>`
  outline: none;
  appearance: none;
  background: white;
  border: none;
  display: flex;

  width: 100%;
  margin: 0;
  height: 58px;
  padding: 16px 24px;
  border-radius: ${({ theme }) => theme.size.borderRadius.br4};
  font-family: ${(props) => props.theme.fonts.FAVORIT};
  font-size: 16px;
  line-height: 24px;
  color: ${(props) => props.theme.color.gray900};
  border: 1px solid ${(props) => props.theme.color.gray300};

  transition: border-color 140ms ease-in;

  &:selected + ${IconWrapper} {
    border: 1px solid ${(props) => props.theme.color.gray900};
  }

  &:hover {
    cursor: pointer;
    border: 1px solid ${(props) => props.theme.color.gray900};
  }

  &:disabled {
    cursor: default;
    background-color: ${(props) => props.theme.color.gray300};
    border: 1px solid ${(props) => props.theme.color.gray300};
    &:hover {
      cursor: default;
    }
  }
`;

const Dropdown: FC<DropdownProps> = ({
  tooltip,
  label,
  options,
  disabled,
  helperText,
  icon,
  register,
  error,
  ...rest
}) => {
  const didMountRef = useRef<boolean>(false);

  useEffect(() => {
    didMountRef.current = true;
  }, []);

  return (
    <Card>
      <TopRow>
        {label && <Label data-testid="dropdown-label">{label}</Label>}
        {!!tooltip && <Tooltip direction="top" text={tooltip} size="small"></Tooltip>}
      </TopRow>
      <Card position="relative">
        <StyledSelect disabled={disabled} {...register} {...rest}>
          {options.map((option) => (
            <option key={option.value} value={option.value}>
              {option.label}
            </option>
          ))}
        </StyledSelect>
        <IconWrapper>{CreateSVGComponent('DownArrow')}</IconWrapper>
      </Card>

      {!!helperText && (
        <Card display="flex" alignItems="center" marginRight="4px">
          {icon && (
            <StyledIcon error={error} $animate={didMountRef.current} marginRight="4px">
              {CreateSVGComponent(error ? 'Error' : 'HelperInfo')}
            </StyledIcon>
          )}
          <HelperText data-testid="text-helper" error={error} $animate={didMountRef.current}>
            {helperText}
          </HelperText>
        </Card>
      )}
    </Card>
  );
};

export default Dropdown;
