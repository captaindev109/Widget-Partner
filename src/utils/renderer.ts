import React from 'react';
import { ConfigProps, Index, EachComponentObjectChildren, BasicAny } from '@/types/types';
import { UIComponents } from './components';

const recursivelyCreateReactComponents = (
  componentToCreate: BasicAny,
  configuration: ConfigProps,
  children: BasicAny
) => {
  if (typeof componentToCreate !== 'undefined') {
    return React.createElement(
      componentToCreate,
      configuration,
      children &&
        (typeof children === 'string'
          ? children
          : children.map((component: EachComponentObjectChildren, index: Index) => viewRenderer(component, index)))
    );
  }
  null;
};

const viewRenderer: BasicAny = (config: ConfigProps, index: Index) => {
  const children = config.children;
  const newConfig = { ...config, key: index };
  const LookForComponentInUILibrary = UIComponents[config.component];
  return recursivelyCreateReactComponents(LookForComponentInUILibrary, newConfig, children);
};

export default viewRenderer;
