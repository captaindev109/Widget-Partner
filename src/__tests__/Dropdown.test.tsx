import { render } from '@testing-library/react';
import renderer from 'react-test-renderer';
import { FC, useState } from 'react';
import { ThemeProvider } from 'styled-components';
import defaultTheme from '@/theme/theme';
import Dropdown, { DropdownOption } from '@/components/Dropdown';

const DROPDOWN_MOCK = [
  {
    label: 'Option 1',
    value: 'option-1',
  },
  {
    label: 'Option 2',
    value: 'option-2',
  },
  {
    label: 'Option 3',
    value: 'option-3',
  },
];

const labelText = "I'm a label";

const DropdownComponent: FC = () => {
  const [option, selectOption] = useState<DropdownOption>({ label: '', value: '' });
  return (
    <ThemeProvider theme={defaultTheme}>
      <Dropdown label={labelText} onChange={(opt) => selectOption(opt)} options={DROPDOWN_MOCK} />
      <span data-testid="test-value">{option.value}</span>
    </ThemeProvider>
  );
};

/** Render tests */
describe('Dropdown render tests', () => {
  test('to see if the component render without crashing', () => {
    render(<DropdownComponent />);
  });
});

/** Snapshot test */
describe('Dropdown snapshot test', () => {
  it('Snapshot test should pass', () => {
    const snap = renderer.create(<DropdownComponent />).toJSON();
    expect(snap).toMatchSnapshot();
  });
});
