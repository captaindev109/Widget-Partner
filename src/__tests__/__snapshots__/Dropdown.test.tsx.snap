// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown snapshot test Snapshot test should pass 1`] = `
Array [
  <div
    className="Card-sc-fb64gk-0 leCJCh"
  >
    <div
      className="Card-sc-fb64gk-0 Dropdown__TopRow-sc-nkn7qa-0 leCJCh ewecwQ"
    >
      <label
        className="Label__StyledText-sc-76lssq-0 eBRjDt"
        data-testid="dropdown-label"
      >
        I'm a label
      </label>
    </div>
    <div
      className="Card-sc-fb64gk-0 iwgbFx"
    >
      <select
        className="Dropdown__StyledSelect-sc-nkn7qa-4 cnoIMZ"
        onChange={[Function]}
      >
        <option
          value="option-1"
        >
          Option 1
        </option>
        <option
          value="option-2"
        >
          Option 2
        </option>
        <option
          value="option-3"
        >
          Option 3
        </option>
      </select>
      <div
        className="Card-sc-fb64gk-0 Dropdown__IconWrapper-sc-nkn7qa-3 leCJCh TcSzt"
      >
        <svg
          fill="none"
          height="11px"
          viewBox="0 0 19 11"
          width="19px"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clipRule="evenodd"
            d="M9.50006 7.9394L2.03039 0.469727L0.969727 1.53039L9.50006 10.0607L18.0304 1.53039L16.9697 0.469727L9.50006 7.9394Z"
            fill="#121212"
            fillRule="evenodd"
          />
        </svg>
      </div>
    </div>
  </div>,
  <span
    data-testid="test-value"
  >
    
  </span>,
]
`;
